default.nix

```diff
@@ -298,7 +315,7 @@ let
         layers = layers;
       };
       fromImageFlag = l.optionalString (fromImage != "") "--from-image ${fromImage}";
-      layerPaths = l.concatMapStringsSep " " (l: l + "/layers.json") ([customizationLayer] ++ layers);
+      layerPaths = l.concatMapStringsSep " " (l: l + "/layers.json") (layers ++ [customizationLayer]);
```

Comment:
The order *does* matter in the case of permissions. If you reverse the above, then the ownership of `/nix/store` ends up with root. From my understanding, this is what is happening:

1. One layer writes to `/nix/store` and puts the owner as 1000:1000
2. Another layer writes to `/nix/store/xxxxxxx` and puts the owner as root. As a result, the ownership of *all* nested paths changes to root. That is to say: `/nix`, `/nix/store/`, and `/nix/store/xxxxxxx` will all have root as the owner.

Since the TAR is being extracted in the order the layers were written, if 1 comes before 2 above, then the desired permissions are preserved.

I've tested this a handful of times myself and confirmed this behavior, but I don't understand why the permissions are changed on all paths as demonstrated above. It seems like a bug, but I don't think it's necessarily happening within `nix2container`, but I could be wrong.

Reply:
Ok, I tried reproducing this with a simplified example and can't do it. So, I suspect I have a bug in my more complicated image. I'll see if I can spend some time whittling it down to see what's going on.

Reply:
nix2container ensures a file can not be added several times with different properties [see here](https://github.com/nlewo/nix2container/blob/686213e7938123a0558737ddc82567e91ecb28db/nix/tar.go#L143)).

However, this check only works for layers built in the same derivation (`buildImage` or `buildLayer`). This means if you build several derivations (for instance `buildImage.layers = [(buildLayer ...)]`), it is possible to override some files. (So, maybe you encounter the issue only when you use the `buildImage.layers` attribute.)

Reply:
Yes, that appears to be the issue. What's weird is that maintaining the change in order seems to actually resolve the issue. For example:

```nix
{
  inputs.nix2container.url = "github:nlewo/nix2container";
  inputs.nixpkgs.follows = "nix2container/nixpkgs";

  outputs = { self, nixpkgs, nix2container }:
    let
      pkgs = import nixpkgs { system = "aarch64-linux"; };
      n2c = nix2container.packages.aarch64-linux.nix2container;
      l = nixpkgs.lib // builtins;

      test = pkgs.runCommand "test" { } ''
        mkdir -p $out/nix/store
      '';
    in
    {
      packages.aarch64-linux.test = n2c.buildImage {
        name = "test";
        tag = "latest";

        copyToRoot = [
          (pkgs.buildEnv {
            name = "root";
            paths = [ pkgs.coreutils pkgs.bashInteractive ];
            pathsToLink = "/bin";
          })
          test
        ];

        layers = [
          (n2c.buildLayer
            {
              deps = [ pkgs.zsh ];
            })
        ];

        perms = [
          {
            path = test;
            regex = "/nix/store";
            uid = 1000;
            gid = 1000;
          }
        ];

        config = {
          entrypoint = [ "${pkgs.bashInteractive}/bin/bash" ];
        };
      };
    };
}
```

If you run the above without the change in this PR, the result is that `/nix/store` is owned by root. If you reverse the order, as done in this PR, the result is that `/nix/store` belongs to uid 1000. I'm not really sure why changing the order makes that result, though...